### ISSUE TITLE
Third-party Dexterity fields not visible in Plone 5.0.X

### ISSUE THREAD

**Issue by zopyx** [plone/Products.CMFPlone#1533]
See

https://community.plone.org/t/custom-dexterity-content-types-show-up-in-plone-4-not-in-plone-5/1957

I could reproduce the problem with another DX 3rd-party field like 

collective.z3cform.datagridfield

The fields just don't show up in Plone 5.

**Comment by zopyx** [comment]
Debugging in 

/home/ajung/.buildout/eggs/plone.schemaeditor-2.0.9-py2.7.egg/plone/schemaeditor/fields.py(62)FieldsVocabularyFactory()

```
(Pdb) for x in field_factories: print x
(u'z3c.relationfield.schema.RelationChoice', <plone.app.relationfield.schemaeditor.RelationFieldFactory object at 0x7f9ff6b80090>)
(u'.xml_binary.XML', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff603c990>)
(u'zope.schema._bootstrapfields.Int', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6d36750>)
(u'zope.schema._bootstrapfields.Bool', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6d367d0>)
(u'plone.app.users.schema.ProtectedTextLine', <plone.app.users.schema.NotEditableFieldFactory object at 0x7f9ff6577450>)
(u'zope.schema._field.Datetime', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6d36850>)
(u'zope.schema._field.Float', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6d36790>)
(u'plone.app.textfield.RichText', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6ccc390>)
(u'plone.namedfile.field.NamedBlobImage', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6cf7290>)
(u'zope.schema._field.Date', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6d36890>)
(u'plone.app.users.schema.ProtectedEmail', <plone.app.users.schema.NotEditableFieldFactory object at 0x7f9ff6577490>)
(u'.xml_field.XML', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff603c190>)
(u'.xml_image.XML', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff5fc5050>)
(u'plone.namedfile.field.NamedBlobFile', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6cf71d0>)
(u'zope.schema._bootstrapfields.TextLine', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6d366d0>)
(u'zope.schema._field.Set', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6db5310>)
(u'zope.schema._field.URI', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff656e350>)
(u'.xpath_field.XMLXPath', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff5fc5ed0>)
(u'plone.schema.email.Email', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff656e390>)
(u'zope.schema._field.Choice', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6d36950>)
(u'zope.schema._bootstrapfields.Text', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6d36710>)
(u'z3c.relationfield.schema.RelationList', <plone.app.relationfield.schemaeditor.RelationFieldFactory object at 0x7f9ff6b80210>)
(u'zope.schema._bootstrapfields.Password', <plone.schemaeditor.fields.FieldFactory object at 0x7f9ff6d36810>)

```

FieldsVocabularyFactory contains my custom 'XML fields. This appears correct.

However the next lines
```
    if context.allowedFields is not None:
        field_factories = [(id, factory) for id, factory in field_factories
                            if id in context.allowedFields]
ajung@dev.zopyx.com:~/.buildout/eggs/plone.schemaeditor-2.0.
```

performs some filtering based on context.allowedFields.


(Pdb) pp context
<TypeSchemaContext at /Plone5/dexterity-types/m1>
(Pdb) pp context.allowedFields
[u'plone.app.textfield.RichText',
 u'plone.namedfile.field.NamedBlobImage',
 u'plone.namedfile.field.NamedBlobFile',
 u'plone.schema.email.Email',
 u'z3c.relationfield.schema.RelationChoice',
 u'z3c.relationfield.schema.RelationList',
 u'zope.schema._bootstrapfields.Bool',
 u'zope.schema._bootstrapfields.Int',
 u'zope.schema._bootstrapfields.Password',
 u'zope.schema._bootstrapfields.Text',
 u'zope.schema._bootstrapfields.TextLine',
 u'zope.schema._field.Choice',
 u'zope.schema._field.Date',
 u'zope.schema._field.Datetime',
 u'zope.schema._field.Float',
 u'zope.schema._field.Set',
 u'zope.schema._field.URI']

So what is the purpose of context.allowedFields? Or how do you enable custom fields for content-types generated through the web?

**Comment by zopyx** [comment]
plone.app.dexterity.types contains a list of hard-coded fields

```
ALLOWED_FIELDS = [
    u'plone.app.textfield.RichText',
    u'plone.namedfile.field.NamedBlobImage',
    u'plone.namedfile.field.NamedBlobFile',
    u'plone.schema.email.Email',
    u'z3c.relationfield.schema.RelationChoice',
    u'z3c.relationfield.schema.RelationList',
    u'zope.schema._bootstrapfields.Bool',
    u'zope.schema._bootstrapfields.Int',
    u'zope.schema._bootstrapfields.Password',
    u'zope.schema._bootstrapfields.Text',
    u'zope.schema._bootstrapfields.TextLine',
    u'zope.schema._field.Choice',
    u'zope.schema._field.Date',
    u'zope.schema._field.Datetime',
    u'zope.schema._field.Float',
    u'zope.schema._field.Set',
    u'zope.schema._field.URI',
]
```

which is assigned to

```
@implementer(ITypeSchemaContext)
class TypeSchemaContext(SchemaContext):

    fti = None
    schemaName = u''
    schemaEditorView = 'fields'
    allowedFields = ALLOWED_FIELDS

```

The code does not seem to take custom fields into account that could be potentially add through the web with the schema editor.

Why is this hard-coded? 
What is the idea not taking custom fields into account?
If it is really hard coded then it is broken.

**Comment by zopyx** [comment]
A possible workaround in Plone 5 is to add customs fields manually to ALLOWED_FIELDS:

```

ALLOWED_FIELDS.append('.xml_field.XML')
ALLOWED_FIELDS.append('.xml_image.XML')
ALLOWED_FIELDS.append('.xml_binary.XML')
```

...however this feels like the wrong way.

**Comment by jensens** [comment]
afaik the addon fields are now available ttw in 5.2
